### PR TITLE
feat: session requests queue

### PR DESCRIPTION
**Pull request by ganchoradkov**
## Description
Implemented `session_request` queue where pending requests are emitted - one after another instead of immediately when received by the SDK

## Type of change

- [x] Chore (non-breaking change that addresses non-functional tasks, maintenance, or code quality improvements)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Draft PR (breaking/non-breaking change which needs more work for having a proper functionality _[Mark this PR as ready to review only when completely ready]_)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How has this been tested?
tests
canary - `2.10.0-canary-af1c2bee`
## Checklist

- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/sign-client/src/controllers/engine.ts [hunk 1]
```diff
@@ -90,6 +99,11 @@ export class Engine extends IEngine {
       this.registerExpirerEvents();
       this.client.core.pairing.register({ methods: Object.keys(ENGINE_RPC_OPTS) });
       this.initialized = true;
+
+      setTimeout(() => {
```
Comment (bkrem) [inline packages/sign-client/src/controllers/engine.ts:103]: Context on why we need a timeout here?

File: packages/sign-client/src/controllers/engine.ts [hunk 2]
```diff
@@ -878,6 +891,43 @@ export class Engine extends IEngine {
     }
   };
 
+  private addRequestToQueue = (request: PendingRequestTypes.Struct) => {
+    this.requestQueue.requests.push(request);
+  };
+
+  private afterResponse = (params: EngineTypes.RespondParams) => {
+    this.deletePendingSessionRequest(params.response.id, { message: "fulfilled", code: 0 });
+    // intentionally delay the emitting of the next pending request a bit
```
Comment (bkrem) [inline packages/sign-client/src/controllers/engine.ts:904]: Why?

File: packages/sign-client/src/controllers/engine.ts [hunk 3]
```diff
@@ -878,6 +891,43 @@ export class Engine extends IEngine {
     }
   };
 
+  private addRequestToQueue = (request: PendingRequestTypes.Struct) => {
+    this.requestQueue.requests.push(request);
+  };
+
+  private afterResponse = (params: EngineTypes.RespondParams) => {
```
Comment (bkrem) [inline packages/sign-client/src/controllers/engine.ts]: Would try to keep naming conventions in line with `<verb>...`

```suggestion
  private cleanupAfterResponse = (params: EngineTypes.RespondParams) => {
```

File: packages/sign-client/src/controllers/engine.ts [hunk 4]
```diff
@@ -878,6 +891,43 @@ export class Engine extends IEngine {
...
+    this.deletePendingSessionRequest(params.response.id, { message: "fulfilled", code: 0 });
+    // intentionally delay the emitting of the next pending request a bit
+    setTimeout(() => {
+      this.requestQueue.state = REQUEST_QUEUE_STATES.idle;
+      this.processRequestQueue();
+    }, toMiliseconds(this.requestQueueDelay));
+  };
+
+  private processRequestQueue = async () => {
+    if (this.requestQueue.state === REQUEST_QUEUE_STATES.active) {
```
Comment (bkrem) [inline packages/sign-client/src/controllers/engine.ts:912]: Q: Trying to fully grok what we're doing here: this is it effectively creating a lock state where queue state remains active until the oldest pending request has been responded to?

File: packages/sign-client/src/controllers/engine.ts [hunk 5]
```diff
@@ -878,6 +891,43 @@ export class Engine extends IEngine {
...
+      this.client.logger.info("session request queue is already active.");
+      return;
+    }
+
+    const request = this.requestQueue.requests[0];
```
Comment (bkrem) [inline packages/sign-client/src/controllers/engine.ts:917]:
```suggestion
    // Select the first/oldest request in the array to ensure last-in-first-out (LIFO)
    const request = this.requestQueue.requests[0];
```

**Review by bkrem — APPROVED** [review]

File: packages/sign-client/src/controllers/engine.ts (hunk 1 above)
Comment (ganchoradkov) [reply packages/sign-client/src/controllers/engine.ts:103]: Timeout is needed to allow wallets to subscribe to `session_request` after `SignClient.init` resolves. The 1s delay is an extra precaution so the first emitted event is not missed due to app not being in a fully ready state to handle events. We can adjust though

File: packages/sign-client/src/controllers/engine.ts (hunk 2 above)
Comment (ganchoradkov) [reply packages/sign-client/src/controllers/engine.ts:904]: same reasoning as above, to allow the app to process the previous request's views/modals/state

File: packages/sign-client/src/controllers/engine.ts (hunk 4 above)
Comment (ganchoradkov) [reply packages/sign-client/src/controllers/engine.ts:912]: Yes, or expires, this way we ensure all requests are responded to

**Pull request merged**